rst_files/orth_proj.rst

```diff
@@ -858,41 +765,46 @@ Exercise 3
 Here's a function that computes the orthonormal vectors using the GS
 algorithm given in the lecture.
 
-.. code-block:: julia
-    :class: Test 
+Setup
+------------------
 
-    using Test # Put this before any code in the lecture. 
+Activate the ``QuantEconLecturePackages`` project environment and package versions
 
-.. code-block:: julia
+.. code-block:: julia 
+
+    using InstantiateFromURL
+    activate_github("QuantEcon/QuantEconLecturePackages")
+    using LinearAlgebra, Statistics, Compat
 
-    using LinearAlgebra
+.. code-block:: julia
+    :class: Test
 
-    function gram_schmidt(X)
+    using Test # Put this before any code in the lecture.
 
-        n, k = size(X)
-        U = zeros(n, k)
+.. code-block:: julia
 
-        # The first col of U is just the normalized first col of X
-        v1 = X[:,1]
-        U[:,1] = v1 / norm(v1)
+    function gram_schmidt(X)
 
-        for i ∈ 2:k
-            # Set up
-            b = X[:,i]        # The vector we're going to project
-            Z = X[:, 1:i-1]   # first i-1 columns of X
+        U = similar(X, Float64)
 
+        function normalized_orthogonal_projection(b, Z)
             # Project onto the orthogonal complement of the col span of Z
-            M = I - Z * inv(Z'Z) * Z'
-            u = M * b
-
+            orthogonal = I - Z * inv(Z'Z) * Z'
+            projection = orthogonal * b
             # Normalize
-            U[:,i] = u / norm(u)
+            output = projection / norm(projection)
```

Comment:
Can we just go `return projection / norm(projection)`?

Reply:
That is a style guide issue. I could suppress the `output =`.

Reply:
OK, let me add to the style guide not to assign a variable name right before returning.

I don't have strong feelings on whether there should be a `return` in circumstances like this.  Sometimes the return makes things clearer for big functions, and sometimes it is completely unnecessar.

Reply:
I think it is a good practice. Stefan for example has pushed for long functions to return `nothing` by default in future releases.

Reply:
Sounds good